src/einsum/lib.py

```diff
@@ -31,45 +33,43 @@ def recurse(pos: List[int], shape: Shape):
     return einsum_tensor(recurse([], shape))
 
 
-@dataclass
-class EinsumOutputSpec:
-
-    # tuple of dimensions (non-negative integers)
-    shape: Shape
+class EinsumOutputSpec(NamedTuple):
 
-    # idxs is a list of integers in [0,52) representing letters and
+    # idxs is a tuple of integers in [0,52) representing letters and
     # of consecutive negative integers from -1 and down representing an ellipsis
     idxs: Idxs
 
-@dataclass
-class EinsumInputSpec:
+class EinsumInputSpec(NamedTuple):
 
     # tuple of dimensions
     shape: Shape
 
-    # idxs is a list of integers in [0,52) representing letters and
+    # idxs is a tuple of integers in [0,52) representing letters and
     # of consecutive negative integers from -1 and down representing an ellipsis
     idxs: Idxs
 
-@dataclass
-class EinsumSpec:
+class EinsumSpec(NamedTuple):
+
+    # inputs is a non-empty tuple of instances of EinsumInputSpec
+    inputs: Tuple[EinsumInputSpec, ...]
+
+    # output is an instance of EinsumOutputSpec
+    output: EinsumOutputSpec
 
     # idxs_map maps indices to dimensions, where an index
     #   is an integer with [0,52) ranging over upper and lower case letters
     #   (e.g. A and Z are 0 and 25, a and z are 26 and 51) and with
     #   negative integers ranging over the indices of the ellipsis, if any
     #   (-1 is the last index of the ellipsis, -2 is the second last, etc)
-    idxs_map: IdxsMap
+    @property
+    def idxs_map(self):
+        return einsum_idxs_map(self.inputs)
 
-    # inputs is a non-empty list of instances of EinsumInputSpec
-    inputs: List[EinsumInputSpec]
+    @property
+    def output_shape(self):
+        return tuple( self.idxs_map[idx] for idx in self.output.idxs )
```

Comment:
it's a good observation that idxs map and output shape can be made property methods since they are derived from inputs and output idxs, and it's neat to not pass in idxs_map to einsum_output(), but on the whole I prefer the old representation which makes it more direct when the computation happens compared to memoization

I'm open to persuasion but, on my first look, my preference is to not change EinsumSpec and just document the redundancy

Reply:
I tend to agree! It also makes it more straightforward to port to C++. Thank you for the thoughtful feedback. :)

One benefit of the current representation is that the einsum specs can be used as keys in a hash map, since everything is a tuple and immutable. Unfortunately dicts and lists are unhashable, so if we store the idx map then this is no longer possible.

I might try a compromise: get rid of the memorization, but keep the tuple representation. That way the idxs map will be computed once for each einsum execution rather than precomputed and stored.

Or I suppose the map could be stored as a tuple of tuples…

(Python’s hash keying is one of the weirder aspects of the language. Lists and dicts really should’ve been hashable, but alas.)

Reply:
yes, changing idxs from list to (immutable) tuple is a good idea

an immutable dict for idxs_map would  be nice (I did a web search and see that others are playing around with that idea, like PEP 416, but couldn't immediately see the best way to go)